markdown/docs/addon-author-guide.md

```diff
@@ -0,0 +1,329 @@
+  init() {
+    componentsToNotify.push(this);
+  },
+
+  windowDidResize() {
+    // ... do some work
+  },
+
+  willDestroy() {
+    for (let i = 0; i < componentsToNotify.length; i++) {
+      if (componentsToNotify[i] === this) {
+        componentsToNotify.splice(i, 1);
+        break;
+      }
+    }
+  }
+});
+```
+
+In this approach, we set up an event listener on the window even if the
+app author never uses our component. It also means that our addon won't
+work in FastBoot, because it tries to access the DOM while defining the
+component.
+
+Instead, we can move this setup into the component's
+`didInsertElement()` hook, which is only invoked once the component is
+inserted into the DOM (i.e., it's only invoked in the browser
+and never in Node).
+
+```js
+// addon/components/resizable-component.js
+import Ember from "ember";
+import $ from "jquery";
+
+let componentsToNotify = [];
+let didSetupListener = false;
+
+function setupListener() {
+  didSetupListener = true;
+  $(window).on('resize', () => {
+    componentsToNotify.forEach(c => c.windowDidResize());
+  });
+}
+
+export default Ember.Component.extend({
+  didInsertElement() {
+    if (!didSetupListener) { setupListener(); }
+    componentsToNotify.push(this);
+  },
+
+  windowDidResize() {
+    // ... do some work
+  },
+
+  willDestroy() {
+    for (let i = 0; i < componentsToNotify.length; i++) {
+      if (componentsToNotify[i] === this) {
+        componentsToNotify.splice(i, 1);
+        break;
+      }
+    }
+  }
+});
+```
+
+This lazy approach moves setup to the first time the component is used,
+improving boot time for routes in the application that don't use this
+component. It also means the component is FastBoot-compatible, because
+all of the code that touches the DOM is contained in the
+`didInsertElement()` hook, which is not invoked in FastBoot.
+
+### Third-Party Dependencies
+
+Many Ember addons wrap other JavaScript libraries in a way that makes
+them easier to use from an Ember app. For example, the
+[ivy-codemirror](https://github.com/IvyApp/ivy-codemirror) addon wraps
+the [CodeMirror code editor](https://codemirror.net/) in a component
+that makes it easy to drop into an Ember app.
+
+Sometimes the library your addon uses is itself incompatible with
+Node.js. When you include the library in your `ember-cli-build.js` file,
+you include code that will prevent the app from running in FastBoot.
+
+If your addon imports third-party code and you are unable to make
+changes to it to add Node compatibility, you can add a guard to your
+`ember-cli-build.js` file to only include it in the browser build.
+
+The FastBoot addon sets the `EMBER_CLI_FASTBOOT` environment variable
+when it is building the FastBoot version of the application. Use
+this to prevent the inclusion of a library in the FastBoot build:
+
+```js
+if (!process.env.EMBER_CLI_FASTBOOT) {
+  // This will only be included in the browser build
+  app.import('some/jquery.plugin.js')
+}
```

Comment:
It may seem obvious, but it's easy to miss if you're just swatting at a "don't break the build" problem -- this means that `jquery.plugin.js` is not available in your app, and it's worth doing a quick check to ensure this is handled gracefully. 

For example, if you're importing d3, and use in helpers for number formatting or something -- the d3 module will now be undefined. The consequence will be that the fastboot-generated HTML will either be incorrect, or incomplete.